pennylane/transforms/decompositions/clifford_t/solovay_kitaev.py

```diff
@@ -0,0 +1,470 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Solovay-Kitaev implementation for approximate single-qubit unitary decomposition."""
+
+import math
+import warnings
+import scipy as sp
+import pennylane as qml
+
+from pennylane.queuing import QueuingManager
+from pennylane.tape import QuantumTape
+from pennylane.transforms.optimization import (
+    cancel_inverses,
+)
+
+# Defining Clifford+T basis
+_CLIFFORD_T_BASIS = {
+    "i": qml.Identity(0),
+    "x": qml.PauliX(0),
+    "y": qml.PauliY(0),
+    "z": qml.PauliZ(0),
+    "h": qml.Hadamard(0),
+    "t": qml.T(0),
+    "tdg": qml.adjoint(qml.T(0)),
+    "s": qml.S(0),
+    "sdg": qml.adjoint(qml.S(0)),
+}
+
+
+class GateSet:
+    """Implements a gate set for storing a sequence of gates and their SU(2) and SO(3) representations
+    for the approximation step in the `Solovay-Kitaev algorithm <https://arxiv.org/abs/quant-ph/0505030>`_.
+
+    Args:
+        operations (Iterable): list of Clifford gates
+    """
+
+    def __init__(self, operations=None):
+        """Instantiate a GateSet object"""
+        self.gates = list(operations) if operations is not None else []
+        self.labels = [op.name for op in self.gates]
+
+        self.matrix = qml.prod(*(self.gates or [qml.Identity(0)])).matrix()
+        self.su2_matrix, self.global_phase = self.get_SU2_matrix(self.matrix)
+        self.so3_matrix = self.get_SO3_matrix(self.su2_matrix)
+
+    @property
+    def name(self):
+        """Return name property for GateSet object"""
+        return "".join(self.labels)
+
+    def __len__(self):
+        """Returns length of gates in GateSet"""
+        return len(self.gates)
+
+    def __getitem__(self, index):
+        """Returns the operation at the given ``index`` in GateSet."""
+        return self.gates[index]
+
+    def __repr__(self):
+        """Returns string representation of GateSet"""
+
+        name = "Gates: ["
+        name += ", ".join(self.labels)
+        name += "], Matrix: " + str(self.matrix)
+        return name
+
+    def __eq__(self, gateset):
+        """Checks equivalence of two GateSet objects"""
+
+        if len(self.gates) != len(gateset.gates):
+            return False
+
+        if self.labels != gateset.labels:
+            return False
+
+        if not qml.math.allclose(self.matrix, gateset.matrix):
+            return False
+
+        if not qml.math.allclose(self.so3_matrix, gateset.so3_matrix):
+            return False
+
+        return True
+
+    @staticmethod
+    def get_SU2_matrix(matrix):
+        """Performs a U(2) to SU(2) transformation via a global phase addition."""
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", RuntimeWarning)
+            factor = qml.math.sqrt((1 + 0j) * qml.math.linalg.det(matrix)) ** -1
```

Comment:
is this the canonical way to ignore complex roots? also, I don't seem to get the warning locally when including the `(1 + 0j)`

Reply:
The warning here is to suppress the warnings raised in `qml.math.linalg.det` (which calls `numpy.linalg.det`). They are ugly-looking `divided by zero` warnings.